embedded: support class existentials with generic classes

eeckstein opened the pull request:
For example:
```
protocol P: AnyObject {}
class C<T> : P {}

let e: P = C<Int>()
```

Generic class existentials are supported in embedded swift by specializing witness tables.
This PR enables the very basic generic class existentials. More work needs to be done to e.g. support protocol inheritance, etc.

Also, I refactored the VTableSpecializer: I moved the VTableSpecializer pass into MandatoryPerformanceOptimizations. MandatoryPerformanceOptimizations already did most of the vtable specialization work.
So it makes sense to remove the VTableSpecializerPass completely and do everything in MandatoryPerformanceOptimizations.

This is a follow-up of https://github.com/swiftlang/swift/pull/76572

eeckstein committed:
SIL: add label to `SubstitutionMap.init`'s bridged argument

eeckstein committed:
SwiftCompilerSources: bridged ProtocolConformance

And use that in some Builder APIs

eeckstein committed:
SIL: add some `Type` APIs and make `Type` a `DiagnosticArgument`

eeckstein committed:
SIL: add some APIs to SubstitutionMap

eeckstein committed:
SIL: improve `VTable` and `WitnessTable`

* add missing APIs
* bridge the entries as values and not as pointers
* add lookup functions in `Context`
* make WitnessTable.Entry.Kind enum cases lower case

eeckstein committed:
SIL: representation for specialized witness tables

The main change here is to associate a witness table with a `ProtocolConformance` instead of a `RootProtocolConformance`.
A `ProtocolConformance` is the base class and can be a `RootProtocolConformance` or a `SpecializedProtocolConformance`.

eeckstein committed:
ModulePassContext: add some utility functions

* `specialize(function:)`
* `deserializeCallees(of:)`
* `createWitnessTable()`
* `createSpecializedVTable`
* `Function.set(isSerialized:)`

eeckstein committed:
embedded: move the VTableSpecializer pass into MandatoryPerformanceOptimizations

MandatoryPerformanceOptimizations already did most of the vtable specialization work.
So it makes sense to remove the VTableSpecializerPass completely and do everything in MandatoryPerformanceOptimizations.

eeckstein committed:
IRGen: make some imported asserts for embedded -> ASSERT to also catch them in NDEBUG compiler builds

eeckstein committed:
IRGen: support emitting specialized witness tables

eeckstein committed:
MandatoryPerformanceOptimizations: specialize witness tables to support class existentials with generic classes

eeckstein commented:
@swift-ci test

eeckstein committed:
SwiftCompilerSources: workaround a host-compiler crash on Windows

Workaround for https://github.com/swiftlang/swift/issues/73253

eeckstein commented:
@swift-ci test

kubamracek commented on include/swift/SILOptimizer/PassManager/Passes.def:
```diff
@@ -258,8 +258,6 @@ SWIFT_FUNCTION_PASS(DeadStoreElimination, "dead-store-elimination",
      "Dead Store Elimination")
 PASS(GenericSpecializer, "generic-specializer",
      "Generic Function Specialization on Static Types")
-PASS(VTableSpecializer, "vtable-specializer",
```
Nice!

kubamracek commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1333,9 +1333,8 @@ void IRGenerator::emitLazyDefinitions() {
     // LazyFunctionDefinitions are allowed, but they must not be generic
     for (SILFunction *f : LazyFunctionDefinitions) {
-      assert(hasValidSignatureForEmbedded(f));
+      ASSERT(hasValidSignatureForEmbedded(f));
```
This is an assert that is active even in production builds, correct? Should I be preferring those in further patches?

eeckstein replied:
Right. Yes, I'm always using this `ASSERT` since we have it, because it's really not good to not catch asserts in an ndebug compiler.

kubamracek commented:
@swift-ci please test

eeckstein committed:
embedded: fix a typo in an error message

eeckstein committed:
SwiftCompilerSources: bridge SILDeclRef

And move NominalTypeDecl and DeclRef from Type.swift into a new file Declarations.swift.
Also add some APIs to NominalTypeDecl

eeckstein commented:
@swift-ci test

eeckstein commented on test/embedded/classes-non-final-method-no-stdlib.swift:
```diff
@@ -3,26 +3,26 @@
 }
 
-func testit3() -> C3<S> {
+public func testit3() -> C3<S> {
```
It's because dead-function-elimination runs before PerfDiags.
But I moved the check back to MandatoryPerformanceOptimizations - and reverted the changes in this test file.

eeckstein commented:
@swift-ci test
